crates/c-api/src/component.rs

```diff
@@ -0,0 +1,659 @@
+        new: wasmtime_component_val_vec_new,
+        empty: wasmtime_component_val_vec_new_empty,
+        uninit: wasmtime_component_val_vec_new_uninitialized,
+        copy: wasmtime_component_val_vec_copy,
+        delete: wasmtime_component_val_vec_delete,
+    )
+    (
+        name: wasmtime_component_val_record_t,
+        ty: wasmtime_component_val_record_field_t,
+        new: wasmtime_component_val_record_field_vec_new,
+        empty: wasmtime_component_val_record_field_vec_new_empty,
+        uninit: wasmtime_component_val_record_field_vec_new_uninitialized,
+        copy: wasmtime_component_val_record_field_vec_copy,
+        delete: wasmtime_component_val_record_field_vec_delete,
+    )
+    (
+        name: wasmtime_component_val_flags_t,
+        ty: u32,
+        new: wasmtime_component_val_flags_new,
+        empty: wasmtime_component_val_flags_new_empty,
+        uninit: wasmtime_component_val_flags_new_uninitialized,
+        copy: wasmtime_component_val_flags_copy,
+        delete: wasmtime_component_val_flags_delete,
+    )
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct wasmtime_component_val_variant_t {
+    pub index: u32,
+    pub val: Option<Box<wasmtime_component_val_t>>,
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct wasmtime_component_val_result_t {
+    pub value: Option<Box<wasmtime_component_val_t>>,
+    pub error: bool,
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct wasmtime_component_val_enum_t {
+    pub discriminant: u32,
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_component_val_flags_set(
+    flags: &mut wasmtime_component_val_flags_t,
+    index: u32,
+    enabled: bool,
+) {
+    let mut f = flags.take();
+    let (idx, bit) = ((index / u32::BITS) as usize, index % u32::BITS);
+    if idx >= f.len() {
+        f.resize(idx + 1, Default::default());
+    }
+    if enabled {
+        f[idx] |= 1 << (bit);
+    } else {
+        f[idx] &= !(1 << (bit));
+    }
+    flags.set_buffer(f);
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_component_val_flags_test(
+    flags: &wasmtime_component_val_flags_t,
+    index: u32,
+) -> bool {
+    let flags = flags.as_slice();
+    let (idx, bit) = ((index / u32::BITS) as usize, index % u32::BITS);
+    flags.get(idx).map(|v| v & (1 << bit) != 0).unwrap_or(false)
+}
+
+#[repr(C, u8)]
+#[derive(Clone)]
+pub enum wasmtime_component_val_t {
+    Bool(bool),
+    S8(i8),
+    U8(u8),
+    S16(i16),
+    U16(u16),
+    S32(i32),
+    U32(u32),
+    S64(i64),
+    U64(u64),
+    F32(f32),
+    F64(f64),
+    Char(char),
+    String(wasmtime_component_val_string_t),
+    List(wasmtime_component_val_vec_t),
+    Record(wasmtime_component_val_record_t),
+    Tuple(wasmtime_component_val_vec_t),
+    Variant(wasmtime_component_val_variant_t),
+    Enum(wasmtime_component_val_enum_t),
+    Option(Option<Box<wasmtime_component_val_t>>),
+    Result(wasmtime_component_val_result_t),
+    Flags(wasmtime_component_val_flags_t),
+}
+
+impl wasmtime_component_val_t {
```

Comment:
Both the to/from conversion from wasmtime_component_val_t <-> component::Val are both inefficient and awkward. I would like to either expose `new_unchecked` methods for individual `Val` subtypes (list, record, etc). These would not do "typechecking", as we're going to have to already do that as we convert. Alternatively, we could upgrade `wasmtime_component_val_t` into the `wasmtime` crate as `component::RawVal` or something and then the conversions could be internal to the crate.